### PR TITLE
Raven db 4818

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: Raven.Database/Bundles/Replication/Controllers/ReplicationController.cs [hunk 1]
```diff
@@ -412,17 +413,21 @@ private void AddSingleConflict(RavenJObject conflict, List<ConflictToResolve> co
...
+        private void SaveReplicationSource(string src, string lastEtag, int batchSize, DateTime lastModified, string collections = null)
         {
             var remoteServerInstanceId = Guid.Parse(GetQueryStringValue("dbid"));
 
             var replicationDocKey = Constants.RavenReplicationSourcesBasePath + "/" + remoteServerInstanceId;
 
             var replicationDocument = Database.Documents.Get(replicationDocKey, null);
             var lastAttachmentId = Etag.Empty;
+            DateTime lastModification = lastModified;
             if (replicationDocument != null)
```
Comment (ayende) [inline Raven.Database/Bundles/Replication/Controllers/ReplicationController.cs:425]: What about deletion? If we have a cluster that is idle for a while, but then we delete a bunch of stuff?

File: Raven.Database/Raft/ClusterManagementHttpClient.cs [hunk 2]
```diff
@@ -210,6 +211,34 @@ public Task SendDatabaseDeleteAsync(string databaseName, bool hardDelete)
             }
         }
 
+        public Task SendReplicationStateAsync(Dictionary<string, Tuple<DateTime,string>> databaseToLastModified)
```
Comment (ayende) [inline Raven.Database/Raft/ClusterManagementHttpClient.cs]: Let us avoid `Tuple<DateTime,string>`. No idea what the string key on the dic is (assume db name), and certainly no clue what the `Item2` in the tuple is meant to be.

File: Raven.Database/Raft/ClusterManager.cs [hunk 3]
```diff
@@ -30,10 +35,89 @@ public class ClusterManager : IDisposable
...
+            {
+                var period = DatabasesLandlord.SystemDatabase.Configuration.Cluster.MaxReplicationLatency;
+                //timer should be null 
+                timer?.Dispose();
+                timer = new Timer(NewLeaderTimerCallback,null, period, period); 
+            }
+            else
+            {
+                //If this node was the leader before, dispose of the timer.
+                timer?.Dispose();
```
Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs]: And set timer to null

File: Raven.Database/Raft/ClusterManager.cs [hunk 4]
```diff
@@ -30,10 +35,89 @@ public class ClusterManager : IDisposable
...
+            var databasesNames = GetDatabasesNames();
+            var databaseToLastModified = new Dictionary<string,Tuple<DateTime,string>>();
+            foreach (var databaseName in databasesNames)
+            {
+                try
+                {
+                    Task<DocumentDatabase> databaseTask;
+                    if (DatabasesLandlord.TryGetOrCreateResourceStore(databaseName, out databaseTask) == false)
+                        continue;
+                    var database = databaseTask.Result;
```
Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs]: You are forcibly loading the databases here, which can take a LONG time.
- Need to protect the timer from concurrent runs.
- Need to skip databases that are currently unloaded
- Need to not trigger loading here (there should be a server startup task that serially load all databases)

File: Raven.Database/Bundles/Replication/Controllers/ReplicationController.cs (hunk 1 above)
Comment (talweiss1982) [inline Raven.Database/Bundles/Replication/Controllers/ReplicationController.cs:425]: we need to get the deletion time from the tombstones too.

File: Raven.Database/Raft/ClusterManager.cs [hunk 5]
```diff
@@ -30,10 +35,89 @@ public class ClusterManager : IDisposable
...
+                    JsonDocument lastDoc = null;
+                    database.TransactionalStorage.Batch(action =>
+                    {
+                        lastDoc = action.Documents.GetDocumentsByReverseUpdateOrder(0, 1).FirstOrDefault();                        
+
+                    });
+                    if (lastDoc == null)
+                    {
+                        if (Log.IsWarnEnabled)
```
Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs]: Not a warning here, just an empty db.
Also, you need to handle deletes as well

File: Raven.Database/Raft/Controllers/ClusterController.cs [hunk 6]
```diff
@@ -144,6 +147,16 @@ public Task<HttpResponseMessage> CanInstallSnapshot([FromUri]CanInstallSnapshotR
             return taskCompletionSource.Task;
         }
 
+
+        [HttpPost]
+        [RavenRoute("cluster/replicationState")]
+        public async Task<HttpResponseMessage> ReplicationState()
```
Comment (ayende) [inline Raven.Database/Raft/Controllers/ClusterController.cs:154]: This is meant for manual triggering of this?

File: Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+        private RaftEngine Engine { get; set; }
+
+        //default should be 2 minutes
+        public TimeSpan MaxReplicationLatency { get; set; }
+
+        public override void Handle(ReplicationStateCommand command)
+        {
+            //If we are the leader we won't have documents sent from ourself so it is a noop.
+            if (Engine.State == RaftEngineState.Leader)
```
Comment (ayende) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: Bad idea, we should have the _exact_ same behavior on all nodes.
You might be processing old commands from the previous leader.

File: Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                return;
+            var lastChanges = command.DatabaseToLastModified;
+            foreach (var databaseName in lastChanges.Keys)
+            {
+                Task<DocumentDatabase> databaseTask;
+                if(Landlord.TryGetOrCreateResourceStore(databaseName, out databaseTask) == false)
+                    throw new InvalidOperationException($"ReplicationStateCommand could not load database: {databaseName}");
+                var database = databaseTask.Result;
```
Comment (ayende) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: You are forcing us to load the database and wait for it, why?
Why aren't we saving this in the cluster state (system db) instead of in each individual db?

We need it for the cluster, not for each db

File: Raven.Database/Server/Controllers/Admin/AdminController.cs [hunk 9]
```diff
@@ -1424,6 +1416,19 @@ private ReplicationTopology CollectReplicationTopology()
             return mergedTopology;
         }
 
+        private HashSet<string> GetDatabasesNames()
```
Comment (ayende) [inline Raven.Database/Server/Controllers/Admin/AdminController.cs:1419]: Duplicate code from `ClusterManager`

File: Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+                var lastModify = lastChanges[databaseName];
+                var docKey = $"{Abstractions.Data.Constants.RavenReplicationSourcesBasePath}/{lastModify.Item2}";
+                var doc = database.Documents.Get(docKey, null);
+                if (doc == null)
```
Comment (talweiss1982) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: I actually need to make sure a document was put to the source at some point, because an idle database with no documents in it will cause this to fail

Comment (ayende) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: Why do you need to do that? Idle db isn't going to shut down unless all its siblings are down as well and there is not traffic.

File: Raven.Database/Raft/ClusterManager.cs [hunk 11]
```diff
@@ -30,10 +35,89 @@ public class ClusterManager : IDisposable
...
+            //Might happen if the timer and the state change event happen at the same time.
+            if (Engine.State != RaftEngineState.Leader)
+            {
+                if(Log.IsDebugEnabled)
+                    Log.Debug($"NewLeaderTimerCallback invoked from non-leader node, actual state:{Engine.State}");
+                return;
+            }
+            var databasesNames = GetDatabasesNames();
+            var databaseToLastModified = new Dictionary<string,Tuple<DateTime,string>>();
+            foreach (var databaseName in databasesNames)
```
Comment (talweiss1982) [inline Raven.Database/Raft/ClusterManager.cs]: Maybe change this to Landlord.ForDatabase

Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs]: You have the `ForAllDatabases` to run something on all active dbs.

What is `ForDatabase` on landlord?

File: Raven.Database/Raft/ClusterManagementHttpClient.cs (hunk 2 above)
Comment (talweiss1982) [inline Raven.Database/Raft/ClusterManagementHttpClient.cs]: it is the transactional id of the source database, it is needed to fetch the source document

Comment (ayende) [inline Raven.Database/Raft/ClusterManagementHttpClient.cs]: So create a class that has a meaningful name so it is readable

File: Raven.Database/Raft/ClusterManager.cs (hunk 4 above)
Comment (talweiss1982) [inline Raven.Database/Raft/ClusterManager.cs]: if the source databse is active the destination is supposed to be active too.
If it is not active it is very well be possible that it is not up to date...

File: Raven.Database/Raft/Controllers/ClusterController.cs (hunk 6 above)
Comment (talweiss1982) [inline Raven.Database/Raft/Controllers/ClusterController.cs:154]: This is how other commands were implemented, i was just keeping it inline with the other commands

File: Raven.Database/Raft/ClusterManager.cs (hunk 4 above)
Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs]: That isn't the issue, the issue is that you are waiting for the db to load, and that shouldn't be here at all

File: Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs (hunk 7 above)
Comment (talweiss1982) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: great, if i was elected and this is the old leader command i want to ignore it.
ignoring it means not failing the append.
if i don't ignore it, i'll fail because of an old leader state.

Comment (ayende) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: You _always_ need to handle it.
The state of all the nodes in the cluster after processing the commands must be the same.
You can't skip it because you are leader or not, that invalidate the whole premise of the state machine.
Leader is transient state.

File: Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs (hunk 8 above)
Comment (talweiss1982) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: the source documents are located at each database...
if the source database is active the destination should be too if not we have a problem anyway

Comment (ayende) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: You can't assume that it is loaded, and you can't block for a long time here.
And you assume state from other servers, that isn't valid

Comment (ayende) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: If the db isn't loaded, you need to start loading it, and skip it.
Then abort the command for retry later.

Another very important thing to remember, there is a very BIG difference between rejecting a command on commit (what you are doing here) and rejecting it on append.

What we need to do is to verify if we can accept the command on append (so failure would skip the log).
What you are doing is running this during commit, after it is in the log.

In other words, even though we failed to commit it, we still have it in the log, and we can still be (very lame) leader.

File: Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs (hunk 10 above)
Comment (talweiss1982) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: database was created in a cluster and nobody wrote to it for 2 minutes

File: Raven.Database/Raft/ClusterManager.cs (hunk 11 above)
Comment (talweiss1982) [inline Raven.Database/Raft/ClusterManager.cs]: i meant that method

File: Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs (hunk 10 above)
Comment (ayende) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: That can't happen if it has siblings, they will ping it every 30 seconds, IIRC. Keeping it up.
That is why we need to start all dbs on startup, it will cause chain reaction around the cluster to keep them up

File: Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs (hunk 7 above)
Comment (talweiss1982) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: this command doesn't change the state besides adding a commit index

File: Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs (hunk 8 above)
Comment (talweiss1982) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: don't we check the commit index for election?

Comment (HibernatingRhinosServiceAccount) [inline Raven.Database/Raft/Storage/Handlers/ReplicationStateCommandHandler.cs]: We just check the last index in the log, not whatever it is committed

File: Raven.Database/Bundles/Replication/Controllers/ReplicationController.cs (hunk 1 above)
Comment (ppekrol) [inline Raven.Database/Bundles/Replication/Controllers/ReplicationController.cs:425]: Tombstones are within the list and everything is ordered by Etag:
https://github.com/ayende/ravendb/blob/v3.5/Raven.Database/Bundles/Replication/Tasks/ReplicationTask.cs#L1284-L1294

Comment (ayende) [inline Raven.Database/Bundles/Replication/Controllers/ReplicationController.cs:425]: Do they have the last modified (which is what we need here)?

Comment (talweiss1982) [inline Raven.Database/Bundles/Replication/Controllers/ReplicationController.cs:425]: yes:
https://github.com/ayende/ravendb/blob/v3.5/Raven.Database/Bundles/Replication/Triggers/VirtualDeleteAndRemoveConflictsTrigger.cs#L62

File: Raven.Database/Raft/ClusterManager.cs [hunk 12]
```diff
@@ -35,14 +37,70 @@ public class ClusterManager : IDisposable
...
+        private readonly TimeSpan maxReplicationLatency;
+        private readonly DateTime clusterManagerStartTime;
+        private void OnProposingCandidacy(object sender, ProposingCandidacyResult e)
+        {
+            var replicationStateDoc = DatabasesLandlord.SystemDatabase.Documents.Get(Constants.Cluster.ClusterReplicationStateDocumentKey,null);
+            if (replicationStateDoc == null)
+            {
+                //This is a case of a node loading for the first time and just never got any replication state.
+                //If we prevent this than a cluster will be non-respnosive when loaded (mostly a test senario but could be a real issue)
+                if (clusterManagerStartTime + maxReplicationLatency + maxReplicationLatency < DateTime.Now)
```
Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs:62]: Use `UtcNow`, not `Now`

File: Raven.Database/Raft/ClusterManager.cs [hunk 13]
```diff
@@ -35,14 +37,70 @@ public class ClusterManager : IDisposable
...
+                {
+                    e.VetoCandidacy = true;
+                    e.Reason = "Could not find replication state document";                    
+                }
+                return;
+            }
+            var replicationState = replicationStateDoc.DataAsJson.ToObject<ReplicationState>();
```
Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs:69]: This wouldn't typically return null, it would throw if there is a serious error, or typically just return empty object.

File: Raven.Database/Raft/ClusterManager.cs [hunk 14]
```diff
@@ -53,71 +111,73 @@ private void OnRaftEngineStateChanged(RaftEngineState state)
...
+                DatabasesLandlord.ForAllDatabases(db => 
+                {
+                    try
                     {
-                        lastDoc = action.Documents.GetDocumentsByReverseUpdateOrder(0, 1).FirstOrDefault();                        
-
-                    });
-                    if (lastDoc == null)
+                        var databaseName = db.Name;
+                        JsonDocument lastDoc = null;
```
Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs:154]: You are assuming here that I'm replicating to the same db name.
Need to verify that this is indeed always the case in clusters (pretty sure it is, mind)

File: Raven.Database/Raft/ClusterManager.cs [hunk 15]
```diff
@@ -234,6 +290,34 @@ public void Dispose()
...
+                if (configuration.DisableReplicationStateChecks == disableReplicationStateChecks)
+                    return;
+                if (configuration.DisableReplicationStateChecks)
+                {
+                    Engine.StateChanged -= OnRaftEngineStateChanged;
+                    SaflyDisposeOfTimer();
+                }
+                else
+                {
+                    Engine.StateChanged += OnRaftEngineStateChanged;
```
Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs:312]: shouldn't you also start the timer here?
If I changed it to false, then to true?

File: Raven.Database/Raft/Dto/ReplicationState.cs [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+        public ReplicationState(Dictionary<string, LastModificationTimeAndTransactionalId> databasesToLastModification)
+        {
+            DatabasesToLastModification = databasesToLastModification;
+        }
+    }
+
+    public class LastModificationTimeAndTransactionalId
+    {
+        public DateTime LastModified { get; set; }
+        public string TransactionalId { get; set; }
```
Comment (ayende) [inline Raven.Database/Raft/Dto/ReplicationState.cs:22]: Use `DatabaseId`, the fact that we internally use the transactional storage id is not relelvant

File: Raven.Database/Raft/ClusterManager.cs (hunk 13 above)
Comment (talweiss1982) [inline Raven.Database/Raft/ClusterManager.cs:69]: i'll change this to try catch

File: Raven.Database/Raft/ClusterManager.cs (hunk 14 above)
Comment (talweiss1982) [inline Raven.Database/Raft/ClusterManager.cs:154]: it is the case

File: Raven.Database/Raft/ClusterManager.cs (hunk 15 above)
Comment (talweiss1982) [inline Raven.Database/Raft/ClusterManager.cs:312]: i need to verify the state of the engine and start the timer only if i'm the leader

Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs:312]: You are doing that all the time, and if you are the leader and the state changed to false and then true, it looks like the timer is disposed and that is it.

File: Raven.Database/Raft/ClusterManager.cs (hunk 13 above)
Comment (ayende) [inline Raven.Database/Raft/ClusterManager.cs:69]: What meaningful thing will you do here?
Also, I've some fixes already, don't do work until I push, please.
